lib/THC/THCTensorMath.cu

```diff
@@ -2270,3 +2189,127 @@ void THCudaTensor_indexSelect(THCState *state, THCudaTensor *res_, THCudaTensor
+             2, "sizes do not match");
+
+  if (!THCudaTensor_pointwiseApply2(state, tensor, mask, TensorMaskedFillOp(value))) {
+    THArgCheck(false, 2, CUTORCH_DIM_WARNING);
+  }
+}
+
+struct TensorMaskedCopyOp {
+  __device__ __forceinline__ void operator()(float* out, float* mask, float* in) {
+    // Really mask should be `0` or `1` but we can't propagate errors here.
+    if (*mask != 0.0f) {
+      *out = *in;
+    }
+  }
+};
+
+void THCudaTensor_maskedCopy(THCState* state,
+                             THCudaTensor *tensor, THCudaTensor *mask, THCudaTensor *src)
+{
+  THArgCheck(THCudaTensor_nElement(state, mask) ==
+             THCudaTensor_nElement(state, src),
+             2, "sizes do not match");
+
+  THCudaTensor_resizeAs(state, tensor, src);
+
+  if (!THCudaTensor_pointwiseApply3(state, tensor, mask, src, TensorMaskedCopyOp())) {
+    THArgCheck(false, 2, CUTORCH_DIM_WARNING);
+  }
+}
+
+struct TensorMaskedSelectOp {
+  TensorMaskedSelectOp(float* t) : out(t) {}
+  __device__ __forceinline__ void operator()(float* mask, float* maskPrefixSum, float* in) {
+    // Really mask should be `0` or `1` but we can't propagate errors here.
+    if (*mask != 0.0f) {
+      out[(int) *maskPrefixSum] = *in;
+    }
+  }
+
+  float* out;
+};
+
+void THCudaTensor_maskedSelect(THCState* state,
+                               THCudaTensor *tensor, THCudaTensor *src, THCudaTensor *mask)
+{
+  THArgCheck(THCudaTensor_nElement(state, mask) == THCudaTensor_nElement(state, src),
+             2, "sizes do not match");
+
+  // Determine our output size
+  THCudaTensor* contigMask = THCudaTensor_newContiguous(state, mask);
+  int totalElements = (int) THCudaTensor_sumall(state, contigMask);
+  THCudaTensor_resize1d(state, tensor, totalElements);
+
+  // Use a prefix sum to determine the output locations of the masked elements
+  THCudaTensor* maskPrefixSum = THCudaTensor_new(state);
+  THCudaTensor_resizeAs(state, maskPrefixSum, mask);
+
+  thrust::device_ptr<float>
+    maskData(THCudaTensor_data(state, contigMask));
+  thrust::device_ptr<float>
+    maskPrefixSumData(THCudaTensor_data(state, maskPrefixSum));
+  thrust::exclusive_scan(maskData,
+                         maskData + THCudaTensor_nElement(state, contigMask),
+                         maskPrefixSumData);
+
+  // Then copy over the masked elements at their desired output index
+  if (!THCudaTensor_pointwiseApply3(
+        state, contigMask, maskPrefixSum,
+        src, TensorMaskedSelectOp(THCudaTensor_data(state, tensor)))) {
+    THArgCheck(false, 2, CUTORCH_DIM_WARNING);
+  }
+
+  THCudaTensor_free(state, contigMask);
+  THCudaTensor_free(state, maskPrefixSum);
+}
+
+void THCudaTensor_maskedFillByte(THCState* state, THCudaTensor *tensor, THByteTensor *mask, float value)
+{
+  THLongStorage* maskSize = THByteTensor_newSizeOf(mask);
+  THCudaTensor* maskCuda = THCudaTensor_newWithSize(state, maskSize, NULL);
+  THLongStorage_free(maskSize);
+
```

Comment:
You need to copy `mask` into `maskCuda`, otherwise it will be uninitialized.
Same in the two functions below.